reconciler/builder.go

```diff
@@ -0,0 +1,133 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright Authors of Cilium
+
+package reconciler
+
+import (
+	"time"
+
+	"github.com/cilium/hive/job"
+	"github.com/cilium/statedb"
+	"github.com/cilium/statedb/index"
+	"golang.org/x/time/rate"
+)
+
+// Register creates a new reconciler and registers it to the application lifecycle.
+//
+// The setStatus etc. functions are passed in as arguments rather than requiring
+// the object to implement them via interface as this allows constructing multiple
+// reconcilers for a single object by having multiple status fields and different
+// functions for manipulating them.
+func Register[Obj comparable](
+	// General dependencies of the reconciler.
+	params Params,
+	// The table to reconcile
+	table statedb.RWTable[Obj],
+
+	// Function for cloning the object.
+	clone func(Obj) Obj,
+
+	// Function for setting the status.
+	setStatus func(Obj, Status) Obj,
+
+	// Function for getting the status.
+	getStatus func(Obj) Status,
+
+	// Reconciliation operations
+	ops Operations[Obj],
+
+	// (Optional) batch operations. Set to nil if not available.
+	batchOps BatchOperations[Obj],
+
+	// zero or more options to override defaults.
+	options ...Option,
+) (Reconciler[Obj], error) {
+	cfg := config[Obj]{
+		Table:           table,
+		GetObjectStatus: getStatus,
+		SetObjectStatus: setStatus,
+		CloneObject:     clone,
+		Operations:      ops,
+		options:         defaultOptions(),
+	}
+	for _, opt := range options {
+		opt(&cfg.options)
+	}
+
+	if cfg.Metrics == nil {
+		if params.DefaultMetrics == nil {
+			cfg.Metrics = NewUnpublishedExpVarMetrics()
+		} else {
+			cfg.Metrics = params.DefaultMetrics
+		}
+	}
+
+	if err := cfg.validate(); err != nil {
+		return nil, err
+	}
+
+	idx := cfg.Table.PrimaryIndexer()
+	objectToKey := func(o any) index.Key {
+		return idx.ObjectToKey(o.(Obj))
+	}
+	r := &reconciler[Obj]{
+		Params:               params,
+		config:               cfg,
+		retries:              newRetries(cfg.RetryBackoffMinDuration, cfg.RetryBackoffMaxDuration, objectToKey),
+		externalPruneTrigger: make(chan struct{}, 1),
+		primaryIndexer:       idx,
+	}
+
+	g := params.Jobs.NewGroup(params.Health)
+
+	g.Add(job.OneShot("reconcile", r.reconcileLoop))
+	if r.config.RefreshInterval > 0 {
+		g.Add(job.OneShot("refresh", r.refreshLoop))
+	}
+	params.Lifecycle.Append(g)
+
+	return r, nil
+}
+
+type Option func(opts *options)
+
+func WithMetrics(m Metrics) Option {
+	return func(opts *options) {
+		opts.Metrics = m
+	}
+}
+
+func WithPruning(interval time.Duration) Option {
+	return func(opts *options) {
+		opts.PruneInterval = interval
+	}
+}
+
+func WithRefreshing(interval time.Duration, limiter *rate.Limiter) Option {
+	return func(opts *options) {
+		opts.RefreshInterval = interval
+		opts.RefreshRateLimiter = limiter
+	}
+}
+
+func WithRetry(minBackoff, maxBackoff time.Duration) Option {
```

Comment:
Should the doc comments from the options move to these public functions, as these are the API now?

Reply:
Yes! Moved.